### ISSUE TITLE
I would like a feature that stops play from working unless I have configured some form of authentication

### ISSUE THREAD

**Issue by geirhe** [mohiva/play-silhouette#291]
See http://cwe.mitre.org/data/definitions/306.html

One of the most common security mistakes out there is to forget to put it in.

I would like to see something that breaks play if an URL isn't configured to be authenticated (and for this purpose, configuring "anonymous access" is an authentication configuration). This would make it much easier to actually use silhouette in an "enterprise" setting.

The opposite applies to basic play applications: if you create a route and a controller, the stuff will work.

I do not have any suggestions for how this should be implemented. I don't know enough about the innards of play. Sorry.

**Comment by akkie** [comment]
In my opinion that would be a task for a static code analysis tool. The problem I see here is that you must  configure the library first to use such functionality. Not all endpoints of an application might be secured. And that makes it as vulnerable as without such functionality.

**Comment by geirhe** [comment]
I can easily create a unit test that verifies that the feature is switched on in my production configuration.
If you are in a work like this (meaning ISO27001/PCI/whatever), you do not have "unsecured" endpoints. Endpoints that should be accessible to all are marked as exactly that. Having an endpoint without a security configuration is a bug.
When doing a security review, it is very nice to be able to show a "security test suite", doing something and showing the auditor that the build chain breaks.

AFAIK, there are no good static code analysis tools for scala.

I definitely think that an authentication library should be designed in such a way that it is hard to use it wrong. This is almost impossible to add consistently after-the-fact by those who did not design the library, no matter how good the tooling around might be. It is a question of domain knowledge.

YMMV.

**Comment by dmcnelis** [comment]
You could also create a trait that extends SecureController and overrides
any Action call to call either UserAwareAction or SecuredAction (depending
on your purposes).  Doing that you could override Action....

Here's the basic idea I'm referring to... we do something similar to pass
around an implicit context.

https://gist.github.com/dmcnelis/23b6d7caeb5873754c61

**Comment by geirhe** [comment]
Exactly. The only thing missing is something that goes and yells if someone tries to use a bare-bones Controller without the security added instead of the one that enforces security.

**Comment by akkie** [comment]
After thinking a bit about the issue, I think such a feature could be implemented with a compiler or SBT plugin, which checks on compile time per reflection if a controller implements a secured or user aware endpoint. Not authenticated endpoints must be explicit annotated with an `Unsecured` annotation.

**Comment by igorbernstein** [comment]
Here's another idea that might easier to implement:

What about introducing a SecuredFilter that a developer can add to GlobalSettings? The filter can run through the authentication process for every request. The biggest change would be to factor out all the authentication logic out of the Silhouette trait so that it can be shared by the Filter & SecuredAction. And perhaps stamping the current authenticator into [a tag](https://github.com/playframework/playframework/commit/1a940282ed70f3f3393766bce94955e24bc074b1) so that the SecuredAction downstream don't have to redo the work of authenticating.

Another Filter that might be useful is an Renewal Filter that can update ildeTimeout on non-secured Actions.

**Comment by geirhe** [comment]
I like it.

**Comment by akkie** [comment]
@igorbernstein You can use a `SecuredAction` or `UserAwareAction` in a filter.

```scala
class SecuredFilter(
  implicit val env: Environment[User, SessionAuthenticator])
  extends Filter with Silhouette[User, SessionAuthenticator] {

  override def apply(next: RequestHeader => Future[Result])(request: RequestHeader): Future[Result] = {
    val action = UserAwareAction.async { r =>
      request.path match {
        case "/admin" if r.identity.isEmpty => Future.successful(Unauthorized)
        case _ => next(request)
      }
    }

    action(request).run
  }
}
```

And then in the Gobal object:
```scala
override def doFilter(a: EssentialAction): EssentialAction = {
  import net.codingwell.scalaguice.InjectorExtensions._
  implicit val env = injector.instance[Environment[User, SessionAuthenticator]]

  Filters(super.doFilter(a), new SecuredFilter)
}
```

**Comment by igorbernstein** [comment]
That's pretty cool! Maybe add that to the docs?

Only thing that's missing is transporting the user/authenticator to the controllers so that they don't have to reauthenticate and more impotantly don't get their authenticator overwritten like in #323

**Comment by akkie** [comment]
> That's pretty cool! Maybe add that to the docs?

http://silhouette.mohiva.com/v2.0/docs/filters

> Only thing that's missing is transporting the user/authenticator to the controllers so that they don't have to reauthenticate

Maybe this could work with request tags as you suggested.

> and more impotantly don't get their authenticator overwritten like in #323

Absolutely, we have to be careful here.

**Comment by akkie** [comment]
So the question here is what is the better solution? A compiler or SBT plugin which does statically analyse the code or the filter approach which might open new issues if we are not careful enough?

**Comment by igorbernstein** [comment]
One doesn't prevent the other....I just think the filter approach can happen a lot sooner...the static analysis has a lot more unknowns

**Comment by wsargent** [comment]
I'd go for the "filter plus tag" approach: it is going to be far less brittle than a compiler / SBT plugin, and allows for more user customization.
